cpp/ql/src/experimental/Security/CWE/CWE-552/DangerousUseRename.ql

```diff
@@ -0,0 +1,66 @@
+/**
+ * @name Dangerous use rename.
+ * @description If the rename function did not work correctly, simply overwriting the destination file can lead to errors.
+ * @kind problem
+ * @id cpp/dangerous-use-rename
+ * @problem.severity warning
+ * @precision medium
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-552
+ */
+
+import cpp
+import semmle.code.cpp.valuenumbering.GlobalValueNumbering
+
+/** Holds if the function argument is used in opening the file for reading. */
+predicate findFileForRead(FunctionCall fc, FunctionCall fc1) {
+  fc1.getTarget().hasGlobalOrStdOrBslName(["open", "fopen"]) and
+  globalValueNumber(fc1.getArgument(0)) = globalValueNumber(fc.getArgument(0)) and
+  fc1.getArgument(1).getValue() = "r" and
+  fc.getASuccessor+() = fc1
+}
+
+/** Holds if the function argument is used in opening the file for writing. */
+predicate findFileForWrite(FunctionCall fc, FunctionCall fc1) {
```

Comment:
Can this be merged with `findFileForWrite` into a predicate that has an extra `mode` argument?

Reply:
Also note that none of this works with Posix `open`, as that takes a numerical `flags` argument. Is that intentional?

Reply:
I'll think about `Posix`. Thanks for the comment.